compiler/passes/buildDefaultFunctions.cpp

```diff
@@ -189,88 +239,74 @@ typedef enum {
   FIND_NOT_REF
 } functionExistsKind;
 
-
 // functionExists returns true iff
 //  function's name matches name
 //  function's number of formals matches numFormals
-//  function's first formal's type matches formalType1 if not NULL
-//  function's second formal's type matches formalType2 if not NULL
-//  function's third formal's type matches formalType3 if not NULL
-static FnSymbol* functionExists(const char* name,
-                                 int numFormals,
-                                 Type* formalType1,
-                                 Type* formalType2,
-                                 Type* formalType3,
-                                 Type* formalType4,
-                                 functionExistsKind kind)
-{
-  switch(numFormals)
-  {
-   default:
-    INT_FATAL("functionExists checks at most 4 argument types.  Add more if needed.");
-    break;
-   case 4:  if (!formalType4)   INT_FATAL("Missing argument formalType4");  break;
-   case 3:  if (!formalType3)   INT_FATAL("Missing argument formalType3");  break;
-   case 2:  if (!formalType2)   INT_FATAL("Missing argument formalType2");  break;
-   case 1:  if (!formalType1)   INT_FATAL("Missing argument formalType1");  break;
-   case 0:  break;
-  }
-
-  const char* nameAstr = astr(name);
-
-  forv_Vec(FnSymbol, fn, gFnSymbols)
-  {
-    if (fn->name != nameAstr)
-      continue;
-
-    // numFormals must match exactly.
-    if (numFormals != fn->numFormals())
-        continue;
-
-    if (formalType1)
-      if (!typeMatch(formalType1, fn->getFormal(1)))
-        continue;
-
-    if (formalType2)
-      if (!typeMatch(formalType2, fn->getFormal(2)))
-        continue;
-
-    if (formalType3)
-      if (!typeMatch(formalType3, fn->getFormal(3)))
+//  function's ith formal type matches formalType[i]
+// We have to template over the vector type here because the cache uses std::vector
+// and gTypeSymbols is a Vec
+template<bool useCache, typename V, size_t numFormals>
+static FnSymbol* functionExists(const char *nameAstr,
+                                const V &fns,
+                                std::array<Type*, numFormals> formalTypes,
+                                functionExistsKind kind) {
+  for (FnSymbol *fn : fns) {
+    if (!useCache) {
+      if (fn->name != nameAstr)
         continue;
+    }
 
-    if (formalType4)
-      if (!typeMatch(formalType4, fn->getFormal(4)))
-        continue;
+    if (numFormals != fn->numFormals())
+      continue;
 
     if (kind == FIND_REF && fn->retTag != RET_REF)
       continue;
 
     if (kind == FIND_NOT_REF && fn->retTag == RET_REF)
       continue;
 
+    size_t i = 0;
+    for_formals(formal, fn) {
+      if (!typeMatch(formalTypes[i], formal))
+        goto next_loop;
```

Comment:
I think it would be better to save a boolean indicating if the function should be returned or something and a `break` statement instead of relying on a goto to get out of the for_formals loop.  That will make it easier to follow